osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs

```diff
@@ -72,49 +55,31 @@ public void Add(InputTrigger trigger)
         protected abstract KeyCounter CreateCounter(InputTrigger trigger);
 
         [BackgroundDependencyLoader]
-        private void load(OsuConfigManager config)
+        private void load(OsuConfigManager config, DrawableRuleset? drawableRuleset)
         {
             config.BindWith(OsuSetting.KeyOverlay, ConfigVisibility);
+
+            if (drawableRuleset != null)
+                AlwaysVisible.BindTo(drawableRuleset.HasReplayLoaded);
         }
 
         protected override void LoadComplete()
         {
             base.LoadComplete();
 
+            controller.OnNewTrigger += Add;
+            AddRange(controller.Triggers);
+
             AlwaysVisible.BindValueChanged(_ => UpdateVisibility());
             ConfigVisibility.BindValueChanged(_ => UpdateVisibility(), true);
         }
 
-        public override bool HandleNonPositionalInput => receptor == null;
-
-        public override bool HandlePositionalInput => receptor == null;
-
-        public partial class Receptor : Drawable
+        protected override void Dispose(bool isDisposing)
         {
-            protected readonly KeyCounterDisplay Target;
-
-            public Receptor(KeyCounterDisplay target)
-            {
-                RelativeSizeAxes = Axes.Both;
-                Depth = float.MinValue;
-                Target = target;
-            }
-
-            public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => true;
-
-            protected override bool Handle(UIEvent e)
-            {
-                switch (e)
-                {
-                    case KeyDownEvent:
-                    case KeyUpEvent:
-                    case MouseDownEvent:
-                    case MouseUpEvent:
-                        return Target.InternalChildren.Any(c => c.TriggerEvent(e));
-                }
-
-                return base.Handle(e);
-            }
+            base.Dispose(isDisposing);
+            controller.OnNewTrigger -= Add;
```

Comment:
When unsubscribing from an event handler on a BDL'd component, you need to `if (controller.IsNotNull())` guard the unsubscription, as there is a possibility that the drawable gets disposal called on it before it even loads (so it will never have dependencies injected into it, thus the nullref death).

Reply:
1a8219ad